### PR TITLE
[SPARK-6229] Add SASL encryption to network library.

### PR DESCRIPTION
**Pull request by vanzin**
There are two main parts of this change:
- Extending the bootstrap mechanism in the network library to add a server-side
  bootstrap (which works a little bit differently than the client-side bootstrap), and
  to allow the  bootstraps to modify the underlying channel.
- Use SASL to encrypt data going through the RPC channel.

The second item requires some non-optimal code to be able to work around the
fact that the outbound path in netty is not thread-safe, and ordering is very important
when encryption is in the picture.

A lot of the changes outside the network/common library are just to adjust to the
changed API for initializing the RPC server.

### CODE REVIEW

**Comment by vanzin** [comment]
I tested this with netty/encryption, netty/no encryption, and nio. Didn't really run performance tests (don't have a cluster available where I can really do that), but I don't expect things with encryption enabled to exactly fly.

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslClientBootstrap.java [hunk 1]
```diff
@@ -62,13 +76,26 @@ public void doBootstrap(TransportClient client) {
...
+        saslClient = null;
+        logger.debug("Channel {} configured for SASL encryption.", client);
+      }
     } finally {
-      try {
-        // Once authentication is complete, the server will trust all remaining communication.
-        saslClient.dispose();
-      } catch (RuntimeException e) {
-        logger.error("Error while disposing SASL client", e);
+      if (saslClient != null) {
```
Comment (squito) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslClientBootstrap.java]: rather than setting `saslClient = null` in the `encrypt` block, and then checking for null here, why not just use `if (!encrypt)` here?  If there is not another reason to `saslClient = null`, I think that would be clearer.

**Comment by squito** [comment]
I'm definitely not an expert on this, but the code seemed OK to me.  My only concern -- is it possible to add a test that ensures data is actually getting encrypting?  Otherwise the unit tests could pass just by making `encrypt = true` a no-op.  I can't think of any way to do that, though -- maybe its not possible.

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslClientBootstrap.java (hunk 1 above)
Comment (vanzin) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslClientBootstrap.java]: Checking for `encrypt` wouldn't work because we still want to dispose of the client instance when some error is thrown.

Comment (squito) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslClientBootstrap.java]: oh, sorry, I should have realized that.  thanks

**Comment by aarondav** [comment]
Will take a look tomorrow, thanks for the ping!

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 2]
```diff
@@ -49,18 +49,21 @@ class NettyBlockTransferService(conf: SparkConf, securityManager: SecurityManage
   private[this] var appId: String = _
 
   override def init(blockDataManager: BlockDataManager): Unit = {
-    val (rpcHandler: RpcHandler, bootstrap: Option[TransportClientBootstrap]) = {
-      val nettyRpcHandler = new NettyBlockRpcServer(serializer, blockDataManager)
-      if (!authEnabled) {
-        (nettyRpcHandler, None)
+    val rpcHandler = new NettyBlockRpcServer(serializer, blockDataManager)
+    val (serverBootstrap: Option[TransportServerBootstrap],
```
Comment (aarondav) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: nit: This syntax gets unwieldy with more code in here, I'd advise switching to something like
```scala
var serverBootstrap: Option[TransportServerBootstrap] = None
var clientBootstrap: Option[TransportClientBootstrap] = None
if (authEnabled) {
  serverBootstrap = Some(new SaslServerBootstrap(transportConf, securityManager)
  clientBootstrap = Some(new SaslClientBootstrap(transportConf, conf.getAppId, securityManager,
    securityManager.isSaslEncryptionEnabled()))
}
```

uses the dreaded `var` but I think it may be less syntactically cumbersome.

File: network/common/src/main/java/org/apache/spark/network/TransportContext.java [hunk 3]
```diff
@@ -55,14 +56,14 @@
   private final Logger logger = LoggerFactory.getLogger(TransportContext.class);
 
   private final TransportConf conf;
-  private final RpcHandler rpcHandler;
+  private final RpcHandler appRpcHandler;
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/TransportContext.java]: Why was this guy renamed? What is the app?

File: network/common/src/main/java/org/apache/spark/network/util/TransportConf.java [hunk 4]
```diff
@@ -105,4 +105,19 @@ public boolean lazyFileDescriptor() {
   public int portMaxRetries() {
     return conf.getInt("spark.port.maxRetries", 16);
   }
+
+  /**
+   * Maximum number of bytes to be encrypted at a time when SASL encryption is enabled.
+   */
+  public int maxSaslEncryptedBlockSize() {
+    return conf.getInt("spark.network.sasl.max_encrypted_block_size_kb", 64) * 1024;
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/util/TransportConf.java]: nit: spark config style is camelCase for things which are not separated by periods

File: network/common/src/main/java/org/apache/spark/network/TransportContext.java (hunk 3 above)
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/TransportContext.java]: Ah -- I see, this is the guy before the bootstraps are executed, which is used by the client directly and as the original handler otherwise.

**Comment by aarondav** [comment]
This is looking very good to me. I've reviewed the core transport part and I like the API. I will have to defer reviewing the rest of the SASL Encryption side and the tests for a bit longer, though, so apologies on that.

**Comment by vanzin** [comment]
A note about the last commit: I noticed while debugging that calling `TransportServer.close()` doesn't close client connections that are still open. That may or may not be a bug in practice, but it could lead to malicious clients hogging resources in case the server process stays alive after the server socket is closed.

That is something for a different change, though. One way to fix it could be to have a timeout on the SASL negotiation: if the negotiation is successful we trust the client, otherwise the timeout closes the door on malicious clients.

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out)
+      throws Exception {
+
+      byte[] data;
+      int offset;
+      int length = msg.readableBytes();
+      if (msg.hasArray()) {
+        data = msg.array();
+        offset = msg.arrayOffset();
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: should we advance msg by the readableBytes?

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+      } while (currentChunk == null && transfered() + written < count());
+
+      transferred += written;
+      return written;
+    }
+
+    private void nextChunk() throws IOException {
+      byteChannel.reset();
+      if (isByteBuf) {
+        int copied = byteChannel.write(buf.nioBuffer());
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: Could this be a ByteArrayOutputStream? If it were, I think you could use buf.readBytes(byteStream) or something of that sort.

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+          currentHeader.skipBytes(bytesWritten);
+          if (currentHeader.readableBytes() > 0) {
+            // Break out of loop if there are still header bytes left to write.
+            break;
+          }
+        }
+
+        target.write(currentChunk);
+        if (!currentChunk.hasRemaining()) {
+          // Only update the count of written bytes once a full chunk has been written.
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: This mechanism may add even more performance burden, as I believe there is special logic which checks for zero-sized writes and backs off from trying to send until the channel appears writable again. However, I can't think of a good way to avoid this that isn't significantly hackier, so it's probably fine.

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslRpcHandler.java [hunk 8]
```diff
@@ -46,19 +47,30 @@
...
+
+  private SparkSaslServer saslServer;
+  private boolean isComplete;
+
+  SaslRpcHandler(
+      TransportConf conf,
+      Channel channel,
+      RpcHandler delegate,
+      SecretKeyHolder secretKeyHolder) {
+    this.conf = conf;
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslRpcHandler.java]: nit: reorder fields to follow same order as constructor parameters

File: network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java [hunk 9]
```diff
@@ -60,13 +60,19 @@
...
+   * QOP value that does not include encryption.
+   */
+  static final String QOP_AUTH = "auth";
+
+  /**
+   * Common SASL config properties for both client and server.
    */
   static final Map<String, String> SASL_PROPS = ImmutableMap.<String, String>builder()
-    .put(Sasl.QOP, "auth")
     .put(Sasl.SERVER_AUTH, "true")
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java]: Is this property relevant for the client? Potentially we could just do away with this static map if not.

File: network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java [hunk 10]
```diff
@@ -60,13 +60,19 @@
   static final String DIGEST = "DIGEST-MD5";
 
   /**
-   * The quality of protection is just "auth". This means that we are doing
-   * authentication only, we are not supporting integrity or privacy protection of the
-   * communication channel after authentication. This could be changed to be configurable
-   * in the future.
+   * QOP value that includes encryption.
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java]: I liked the spelled-out "quality of protection" better :)

File: network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java [hunk 11]
```diff
@@ -75,11 +81,20 @@
...
 
-  public SparkSaslServer(String secretKeyId, SecretKeyHolder secretKeyHolder) {
+  public SparkSaslServer(
+      String secretKeyId,
+      SecretKeyHolder secretKeyHolder,
+      boolean alwaysEncrypt) {
     this.secretKeyId = secretKeyId;
     this.secretKeyHolder = secretKeyHolder;
+
+    String qop = alwaysEncrypt ? QOP_AUTH_CONF : String.format("%s,%s", QOP_AUTH_CONF, QOP_AUTH);
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java]: I assume this is a comma-separated list of the supported formats, for negotiation? Maybe add a comment to this effect.

File: network/common/src/test/resources/log4j.properties [hunk 12]
```diff
@@ -23,5 +23,5 @@ log4j.appender.file.file=target/unit-tests.log
 log4j.appender.file.layout=org.apache.log4j.PatternLayout
 log4j.appender.file.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss.SSS} %t %p %c{1}: %m%n
 
-# Silence verbose logs from 3rd-party libraries.
+# Filter debug messages from noisy 3rd-party libs.
```
Comment (aarondav) [inline network/common/src/test/resources/log4j.properties]: kind of a funny change

File: network/shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java [hunk 13]
```diff
@@ -58,10 +60,15 @@
   public ExternalShuffleClient(
       TransportConf conf,
       SecretKeyHolder secretKeyHolder,
-      boolean saslEnabled) {
+      boolean saslEnabled,
+      boolean saslEncryptionEnabled) {
+    Preconditions.checkArgument(
+        !saslEncryptionEnabled || saslEnabled,
```
Comment (aarondav) [inline network/shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java]: nit: I think 2 space indent is appropriate here

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.FileRegion;
+import io.netty.handler.codec.MessageToMessageDecoder;
+import io.netty.util.AbstractReferenceCounted;
+import io.netty.util.ReferenceCountUtil;
+
+import org.apache.spark.network.util.ByteArrayWritableChannel;
+import org.apache.spark.network.util.NettyUtils;
+
+class SaslEncryption {
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: Please add a class comment

File: network/common/src/main/java/org/apache/spark/network/TransportContext.java [hunk 15]
```diff
@@ -99,9 +108,9 @@ public TransportServer createServer() {
    * be used to communicate on this channel. The TransportClient is directly associated with a
    * ChannelHandler to ensure all users of the same channel get the same TransportClient object.
    */
-  public TransportChannelHandler initializePipeline(SocketChannel channel) {
+  public TransportChannelHandler initializePipeline(SocketChannel channel, RpcHandler rpcHandler) {
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/TransportContext.java]: The rpcHandler vs appRpcHandler thing is definitely confusing to someone reading this for the first time, please add a comment here or in TransportContext about the difference.

File: network/common/src/test/java/org/apache/spark/network/sasl/SparkSaslSuite.java [hunk 16]
```diff
@@ -86,4 +117,237 @@ public void testNonMatching() {
...
     }
   }
+
+  @Test
+  public void testSaslAuthentication() throws Exception {
+    testBasicSasl(false);
+  }
+
+  @Test
+  public void testSaslEncryption() throws Exception {
```
Comment (aarondav) [inline network/common/src/test/java/org/apache/spark/network/sasl/SparkSaslSuite.java]: I think these methods should be right next to testBasicSasl's definition

**Comment by aarondav** [comment]
LGTM, only minor comments. The tests look good. Apologies for taking so long to review!

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java (hunk 5 above)
Comment (vanzin) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: It's unnecessary since `MessageToMessageDecoder` will release the input message when this method returns.

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java (hunk 6 above)
Comment (vanzin) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: Hmm... I could, but then I'd need different logic for the `FileRegion` case (since it needs a `WritableByteChannel`).

It's not a lot of code, but is more than just leaving the `skipBytes` call there.

(I could also implement `GatheringByteChannel` instead of `WritableByteChannel`, but, again, that would be more code than just the one skip call...)

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java (hunk 7 above)
Comment (vanzin) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: I see. I could return `1` instead, and keep the counts synchronized, but some pathological case might still require returning `0`. That should be rare, though, so let me do this to avoid the performance hit in the common case.

File: network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java (hunk 9 above)
Comment (vanzin) [inline network/common/src/main/java/org/apache/spark/network/sasl/SparkSaslServer.java]: I don't think it applies to the client. I'm also not sure whether it's needed at all, but I'll change the code so it's only set for the server.

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java (hunk 5 above)
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: I see, it's just slightly odd that only one of the two cases moves msg's reader index. In the current implementation, you're right that it looks correct, but a seemingly unrelated change could cause issues down the line due to the discrepancy, and only in a very weird circumstance (off-heap vs on-heap messages). Consider using `msg.getBytes(msg.readerIndx(), data)` below.

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java (hunk 6 above)
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: I think `Channels.newChannel(outputStream)` might work.

(The reason I'd especially like this solution is that we could keep ByteArrayWritableChannel in src/test if we could just resolve this use-case without it.)

**Comment by vanzin** [comment]
I made the changes to report non-zero written bytes (https://github.com/vanzin/spark/commit/346f8291a91b0b73742263d7a23ee4027d76adb9). But after writing and testing it, I kinda started questioning whether it will improve anything...

`0` or `1` are basically returned in the same situation: the underlying channel does not have enough buffer. So returning non-zero would cause netty to "busy loop" and call `transferTo` again, which would fail to write anything and thus would now return 0.

But I guess this makes the code behave more like it would with other non-hacky `FileRegion` implementations.

**Comment by vanzin** [comment]
> Can probably eliminate this by constructing it in nextChunk(), I believe that's the only place it's used

That would generate more garbage (a new instance per call to `nextChunk`). Probably not a big deal, though...

File: network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+     *
+     * This makes assumptions about how netty treats FileRegion instances, because there's no way
+     * to know beforehand what will be the size of the encrypted message. Namely, it assumes
+     * that netty will try to transfer data from this message while
+     * <code>transfered() < count()</code>. So these two methods return, technically, wrong data,
+     * but netty doesn't know better.
+     */
+    @Override
+    public long count() {
+      return isByteBuf ? buf.readableBytes() : region.count();
```
Comment (aarondav) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: won't this count() change as we read bytes?

Comment (vanzin) [inline network/common/src/main/java/org/apache/spark/network/sasl/SaslEncryption.java]: It's unclear from the docs, but in the existing implementations (e.g. LazyFileRegion), `count()` always returns the same value. Also, netty does [this](https://github.com/netty/netty/blob/33f75d374091946058fc334c3cdbcd0f0a59d9b3/transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java#L189):

    boolean done = region.transfered() >= region.count();

Which implies that `count()` should not change as bytes are transferred.

**Comment by rxin** [comment]
@vanzin Give me a day or two. I'm checking with Netty guys on the overall design pattern. I'm worried about CPU heavy encryption blocking the IO thread calling transferTo. One solution might be to merge this, and follow up with a big TODO and JIRA to improve it.

**Comment by vanzin** [comment]
>  I'm worried about CPU heavy encryption blocking the IO thread calling transferTo

Yeah, that's the part that sucks with the current approach. But I couldn't find an alternative.

- ThreadPerChannelEventLoop can solve the problem but doesn't sound like the right approach.
- Using a separate thread pool for the encryption / decryption handlers is possible, but we still need to guarantee ordering when pushing messages on the outbound path; I don't think you're allowed to call blocking methods (like `ChannelPromise.sync()`) in channel threads.

Alternative suggestions are mostly welcome, if available.

**Comment by rxin** [comment]
Thanks. I'm merging this in master.

**Comment by JoshRosen** [comment]
How do we enable this feature? Would be good to add some docs for this.
